### ISSUE TITLE
--save-quarantine can overwrite test files

### ISSUE THREAD

**Issue by okken** [EnergySage/pytest-quarantine#29]
This works great: `pytest test_foo.py --save-quarantine`
This overwrites the test file: `pytest --save-quarantine test_foo.py`
:(

**Comment by bhrutledge** [comment]
Yikes, thanks for reportng. That's an issue with `--save-quarantine` optionally accepting a filename (instead of the default `quarantine.txt`):

```
quarantine:
  --save-quarantine=[PATH]
                        Write failing tests to PATH (default: quarantine.txt)
  --quarantine=[PATH]   Mark tests listed in PATH with `xfail` (default:
                        quarantine.txt)
```

The workaround would be `pytest --save-quarantine -- test_foo.py`. That said, I've been on the fence about this whole "optional path" thing. One resolution would be to add a dedicated option, such as `--quarantine-path=PATH`. I'm open to other suggestions.

Also, FWIW, I've only ever used `--save-quarantine` when running the whole test suite. I've found it more common to run a specific test module with the quarantine, e.g. `pytest --quarantine -- test_foo.py`.

**Comment by bhrutledge** [comment]
Another option would be to make the file path required, e.g.

```
quarantine:
  --save-quarantine=PATH   Write failing tests to PATH
  --quarantine=PATH        Mark tests listed in PATH with `xfail`
```

And you'd always run:

```
$ pytest --save-quarantine=quarantine.txt
$ pytest --quarantine=quarantine.txt
```

I was attempting to keep the command short by providing a reasonable default, but it does add some complexity. I also note that other plugins that write files require an argument:

```
$ pytest --junit-xml=results.xml --html=results.html --result-log=results.txt
```

I'm inclined to take this approach. @okken what do you think?

**Comment by okken** [comment]
I was thinking along the lines of checking the path. If it’s directory, fail. If it’s a .py file, fail. Not sure what the format of your file is, but it is either going to be a nonexistent file, an empty file, or a file that you could detect is a quarantine file. If not, bail with a usage warning error. 

Just an option. 

I think keeping the simple case simple is important.

**Comment by bhrutledge** [comment]
Thanks for the suggestion. That sounds like a good way to fix the bug associated with the current command line behavior. For `--save-quarantine`, I think that would mean attempting to open and parse the quarantine file (which is just a list of test ID's like `tests/test_foo.py::test_bar`) prior to the test run.

However, I'm still leaning towards requiring an explicit path on the command line, because, ya know, that's [better than implicit](https://www.python.org/dev/peps/pep-0020/#the-zen-of-python). 😉  I think it will make plugin's code, tests, and documentation simpler by removing conditional clauses around the file path.

That said, your suggestion has me thinking that a useful enhancement would be to:

- Check that we can write the quarantine when using `--save-quarantine`
- Check that the quarantine contains valid test ID's when using `--quarantine`

These checks would be done before the tests run, because the test suite can take a long time, and it'd be a bummer to have to run it again because of an invalid quarantine file.

**Comment by okken** [comment]
Sounds good.

**Comment by bhrutledge** [comment]
@okken FYI, I released this in version 2.0.0a0. I'm hoping to release 2.0.0 within a week (or so) to resolve the other open issues.
